### DISCUSSION TITLE
Steps to use kyber768 KEM in code

### DISCUSSION THREAD

**Discussion by ranjankumar23** [open-quantum-safe/openssl#461] [Q&A]
I want to know steps to use kyber768 KEM in SSL communication.

I already have a code which is doing SSL communication using openssl APIs like SSL_new, SSL_connect, SSL_accept etc.
I compiled quantum safe oqs and quantum safe openssl. 

1. Now to use **_kyber768 KEM_** i only have to add this KEM algo using this API SSL_CTX_set1_groups_list?
2. Or i have to call any other API like this?
`    testresult &=
      (ctx = EVP_PKEY_CTX_new_from_name(libctx, kemalg_name, NULL)) != NULL
      && EVP_PKEY_keygen_init(ctx)
      && EVP_PKEY_generate(ctx, &key);
`

**Comment by dstebila** [comment]
Are you using OQS-OpenSSL 1.1.1 or the OQS Provider with OpenSSL 3?  This question would be better asked in the appropriate repository.

**Reply by ranjankumar23** [reply]
I am using OQS-OpenSSL 1.1.1

**Reply by baentsch** [reply]
`EVP_PKEY_CTX_new_from_name` is an OpenSSL3 API not present in oqs-openssl111. You can't possibly use that. `SSL_CTX_set1_groups_list` in turn is an API already available in 111.

**Reply by ranjankumar23** [reply]
@baentsch Thanks for your reply.

Please confirm if SSL client and server just use this API SSL_CTX_set1_groups_list to add kyber768 as KEM.

**Reply by ranjankumar23** [reply]
@baentsch Michael, not sure if my question make sense.
I am in assumption that key exchange kyber768 is used during SSL connection if SSL client and server use this API SSL_CTX_set1_groups_list to add kyber768 as KEM.

**Reply by baentsch** [reply]
That's just one way to do it. AFAIK you could also do it via config file. Please consult the OpenSSL documentation -- considering any PQ KEM as a KEM/"group" in OpenSSL terminology. Sorry if I cannot confirm anything for OpenSSL111: This code is very old and I'm not familiar with it.

**Reply by ranjankumar23** [reply]
Thanks @baentsch .
This 'OQS-OpenSSL_1_1_1' is the latest branch of open quantum safe openssl.
1. Do we have any QOS openssl branch based on openssl3.0?
2. and i am planning to use https://github.com/open-quantum-safe/oqs-demos/blob/main/wireshark/USAGE.md#quick-start to verify between client and server.

**Reply by baentsch** [reply]
> This 'OQS-OpenSSL_1_1_1' is the latest branch of open quantum safe openssl.

Well, latest, last and final :-)

> Do we have any QOS openssl branch based on openssl3.0?

Well, Yes (OQS-OpenSSL3) but this is discontinued as it's pointless: The code there contains the first steps creating `oqs-provider` (now a separate project) and `oqs-provider` brings all the same (and more) PQ features into OpenSSL3 (without changing OpenSSL3). But I think you know that already.

> and i am planning to use https://github.com/open-quantum-safe/oqs-demos/blob/main/wireshark/USAGE.md#quick-start to verify between client and server.

You can do that: That is completely independent of any client- and server software.

**Reply by ranjankumar23** [reply]
Understood @baentsch previous reply was helpful.

I am able to verify kyber768 using openquantumsafe test server.
sudo docker run --net=host --privileged --env="DISPLAY" --volume="$HOME/.Xauthority:/root/.Xauthority:rw" openquantumsafe/wireshark
sudo docker run -v `pwd`:/home/build/ranjan -it openquantumsafe/curl:0.8.0 curl --cacert /home/build/ranjan/CA.crt https://test.openquantumsafe.org:6015 --curves kyber768

But same wireshark is not showing TLS traffic with the actual quantum-safe KEM algorithm between client and server running on my local machine. my client server is using SSL_CTX_set1_groups_list to add kyber768.

**Reply by baentsch** [reply]
> But same wireshark is not showing TLS traffic with the actual quantum-safe KEM algorithm between client and server running on my local machine. 

´wireshark´ is a tool to trace network traffic. Client and server on the same machine generate no network traffic.

**Reply by dstebila** [reply]
> > But same wireshark is not showing TLS traffic with the actual quantum-safe KEM algorithm between client and server running on my local machine. 
> 
> 
> 
> ´wireshark´ is a tool to trace network traffic. Client and server on the same machine generate no network traffic.
> 
> 
You can run wireshark on the loopback interface to collect local traffic, if memory serves.

**Reply by ranjankumar23** [reply]
Thank you @dstebila and @baentsch. In my application I am not getting client and server hello handsake packets and also it is showing TLS2 Application.....as description of packets. I will share network capture. But do you see any problem with my application?

Also I want to print kem.c under kyber768 case. But I don't see my prints even when running s_client.

**Reply by ranjankumar23** [reply]
1. TLS1.2 issue mentioned above is my application issue.
2.  I need your help into how I can see kem.c logs/prints.

**Reply by ranjankumar23** [reply]
I am able to see key exchange mechanism is using kyber768 in my application.
I have one question there---------Why TLS version is 1.2 inside but outside it is TLS1.3 as shown in attachment?
![share](https://github.com/open-quantum-safe/openssl/assets/5661396/92fa236b-6cab-41e3-bc05-2d1ee8e6992c)

**Reply by dstebila** [reply]
The TLS 1.3 spec ([RFC 8446](https://www.rfc-editor.org/rfc/rfc8446), page 28) works this way.

**Reply by ranjankumar23** [reply]
Thank you @dstebila . We can close this thread.
